### PR TITLE
Make LRU caching tree-based so subtrees of the cache can be invalidated cheaply.

### PR DESCRIPTION
**Pull request by dbkr**
Use this to invalidate the event_push_actions cache appropriately and fix bug where the unread / notify count just wouldn't update.

### CODE REVIEW

File: synapse/util/caches/treecache.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+SENTINEL = object()
+
+
+class TreeCache(object):
```
Comment (erikjohnston) [inline synapse/util/caches/treecache.py]: Comments please.

- What does this do?
- What assumptions are there on the key/values?

**Comment by erikjohnston** [comment]
`TreeCache` does not seem implement `__contains__` which is required by `LruCache`

File: synapse/util/caches/lrucache.py [hunk 2]
```diff
@@ -17,11 +17,23 @@
...
+                yield m
+
 
 class LruCache(object):
     """Least-recently-used cache."""
-    def __init__(self, max_size):
-        cache = {}
+    def __init__(self, max_size, keylen):
+        cache = TreeCache()
+        self.size = 0
```
Comment (erikjohnston) [inline synapse/util/caches/lrucache.py]: Why is this on `self`? My understanding was that everything here is explicitly *not* attached to `self` for performance reasons.

**Comment by erikjohnston** [comment]
I'd have preferred that the `LruCache` did not default to using the `TreeCache` without knowing what the performance characteristics are of `TreeCache` vs `dict`. We use `LruCache` in a number of criticial places, including the event cache and state cache.

If we do default `LruCache` to using `TreeCache`, please document this fact on `LruCache` and that they keys _must_ be tuples.

**Comment by erikjohnston** [comment]
Does the `TreeCache` ever get pruned? Or will it fill up with empty dicts?

**Comment by NegativeMjark** [comment]
> Does the TreeCache ever get pruned? Or will it fill up with empty dicts?

The last bit of the pop method prunes the empty dicts.

File: synapse/util/caches/lrucache.py [hunk 3]
```diff
@@ -123,6 +150,7 @@ def cache_contains(key):
         self.set = cache_set
         self.setdefault = cache_set_default
         self.pop = cache_pop
+        self.del_multi = cache_del_multi
```
Comment (erikjohnston) [inline synapse/util/caches/lrucache.py]: Might be nice to only define this if `cache` is a `TreeCache`?

File: synapse/util/caches/lrucache.py (hunk 2 above)
Comment (dbkr) [inline synapse/util/caches/lrucache.py]: capturing is not so straightforward for things that need to be mutable :/

**Comment by dbkr** [comment]
ptal

**Comment by erikjohnston** [comment]
LGTM

**Pull request merged**